src/universe/visualization/img_tools.py

```diff
@@ -14,3 +19,352 @@ def rotate(img: np.ndarray, angle: float) -> np.ndarray:
+               line_width: float = 0.004, opacity: float = 0.75) -> np.ndarray:
+    """
+    Draws the given lines on the given image with the given color.
+    Make sure that the lines are scaled to the image size.
+
+    :param image: image to draw the lines on
+    :param lines: lines to draw on the image (scaled to the image size)
+    :param color: color of the lines (in the same color space as the image)
+    :param endpoints: whether to draw circles at the endpoints of the lines
+    :param line_width: this value multiplied by the width of the image is the line width in pixels
+    :param opacity: opacity of the lines (0-100)
+    :return: image with the lines drawn on it
+    """
+    if image is None:
+        raise ValueError('Image does not exist')
+    if not (0. <= opacity <= 1.):
+        print('Opacity percentage must be between 0 and 1')
+        opacity = min(max(0., opacity), 1.)
+
+    # return the image if there are no lines to draw
+    if not lines:
+        return image
+
+    # calculate the line width in pixels based on the image width
+    line_width = round(line_width * image.shape[1])
+
+    # create an overlay for the lines
+    overlay = np.zeros_like(image)
+    for line in lines:
+        # round the points to integers
+        p1 = (round(line.a.x), round(line.a.y))
+        p2 = (round(line.b.x), round(line.b.y))
+
+        # draw line on the overlay
+        cv.line(overlay, p1, p2, color, line_width)
+        # draw endpoints on the overlay
+        if endpoints:
+            cv.circle(overlay, p1, line_width, color, -1)
+            cv.circle(overlay, p2, line_width, color, -1)
+
+    # draw the overlay on the image
+    return cv.addWeighted(overlay, opacity, image, 1., 0.)
+
+
+def draw_mask(image: np.ndarray, mask: np.ndarray, color: Tuple[int, int, int],
+              threshold: int = 30, opacity: float = 0.75) -> np.ndarray:
+    """
+    Draws the given mask on the given image.
+
+    :param image: image to draw the mask on
+    :param mask: mask to draw
+    :param color: color of the mask (in the same color space as the image)
+    :param threshold: threshold of the mask (0-255)
+    :param opacity: opacity of the mask (0-1)
+    :return: image with the mask drawn on it
+    """
+    if image is None:
+        raise ValueError('Image does not exist')
+    if not (0. <= opacity <= 1.):
+        print('Opacity percentage must be between 0 and 1')
+        opacity = min(max(0., opacity), 1.)
+
+    # return the image if there is no mask to draw
+    if mask is None:
+        return image
+
+    # resize the mask
+    mask = cv.resize(mask, (image.shape[1], image.shape[0]))
+    # convert the mask to grayscale
+    mask = cv.cvtColor(mask, cv.COLOR_RGB2GRAY)
+    # threshold the mask
+    mask = np.where(mask < threshold, 0, 1)
+    # create overlay
+    overlay = np.zeros_like(image)
+    # draw mask on the overlay
+    overlay[mask > 0] = color
```

Comment:
```suggestion
    overlay[mask] = color
```
Since mask is a matrix of 0s and 1s, it can be used as a bool matrix: (mask > 0) = mask.astype(bool)

Reply:
This freezes the program